src/components/tl/cuda/bcast/bcast_linear.c

```diff
@@ -0,0 +1,423 @@
+static inline ucc_status_t peer_find_free_barrier(ucc_tl_cuda_task_t *task)
+{
+    ucc_tl_cuda_team_t *team = TASK_TEAM(task);
+    uint32_t max_concurrent  = UCC_TL_CUDA_TEAM_LIB(team)->cfg.max_concurrent;
+    ucc_tl_cuda_shm_barrier_t *curr_bar;
+    int                        i;
+    ucc_status_t               st;
+
+    for (i = 0; i < max_concurrent; ++i) {
+        curr_bar = UCC_TL_CUDA_TEAM_BARRIER(team, max_concurrent + i);
+        // Check if the barrier is claimed by the task's root
+        if (curr_bar->tag == task->bcast_linear.key) {
+            task->bar = curr_bar;
+            st        = ucc_tl_cuda_shm_barrier_init_root(
+                task->subset.map.ep_num, task->subset.myrank,
+                task->bcast_linear.root, task->bar);
+            if (ucc_unlikely(st != UCC_OK)) {
+                ucc_error("failed to init peer barrier");
+                return UCC_ERR_NO_RESOURCE;
+            }
+            task->coll_id = i + max_concurrent;
+            return UCC_OK;
+        }
+    }
+    // try next time
+    return UCC_ERR_NOT_FOUND;
+}
+
+static ucc_status_t ucc_tl_cuda_bcast_linear_finalize(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_cuda_task_t *task = ucc_derived_of(coll_task, ucc_tl_cuda_task_t);
+
+    tl_trace(UCC_TASK_LIB(task), "finalizing task %p", task);
+    ucc_tl_cuda_task_put(task);
+    return UCC_OK;
+}
+
+static void ucc_tl_cuda_bcast_linear_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_cuda_task_t *task              = ucc_derived_of(coll_task, ucc_tl_cuda_task_t);
+    ucc_tl_cuda_team_t *team              = TASK_TEAM(task);
+    ucc_rank_t          trank             = UCC_TL_TEAM_RANK(team);
+    size_t              half_scratch_size = get_raw_scratch_size(team) / 2;
+    ucc_rank_t          tsize             = UCC_COLL_ARGS_ACTIVE_SET(&TASK_ARGS(task))
+                                            ? (ucc_rank_t)task->subset.map.ep_num
+                                            : UCC_TL_TEAM_SIZE(team);
+    size_t              chunk_size        =
+                            task->bcast_linear.step < task->bcast_linear.num_steps
+                            ? ucc_min(half_scratch_size, task->bcast_linear.size)
+                            : task->bcast_linear.size -
+                                (task->bcast_linear.step - 1) * half_scratch_size;
+    size_t              offset_buff       = task->bcast_linear.step * half_scratch_size;
+    ucc_ee_executor_t         *exec;
+    ucc_ee_executor_task_t    *etask;
+    void                      *sbuf, *dbuf;
+    ucc_rank_t                 peer;
+    ucc_status_t               st;
+    int                        i;
+
+    task->super.status = UCC_INPROGRESS;
+
+    st = ucc_coll_task_get_executor(&task->super, &exec);
+    if (ucc_unlikely(st != UCC_OK)) {
+        task->super.status = st;
+        return;
+    }
+
+    switch (task->bcast_linear.stage) {
+    case STAGE_INIT_BAR_ROOT:
```

Comment:
suggesting, not requesting: `STAGE_INIT_BAR_ROOT` and `STAGE_FIND_BAR_PEER` could be merged into one stage, with a if/else inside. This has the benefit of 1) reducing the number of cases, and, 2) allowing to fall through the next case `STAGE_SYNC` and thus saving some latency